Update MATCH page (#1080)
neo4j/docs-cypher: modules/ROOT/pages/clauses/match.adoc

:description: The `MATCH` clause is used to search for the pattern described in it.
include::https://raw.githubusercontent.com/neo4j-graphacademy/courses/main/asciidoc/ads/data-analysis.adoc[]

= MATCH

> The `MATCH` clause enables you to define specific patterns that the database will search for within its graph structure.

Reviewer: general note about "will" - i think there's a style guide rule to avoid it where possible. there are many occurrences in this page, but i can also see them being warranted because a query happens at a certain point in time and then the database operations described happen later. i guess it comes down to personal preference :)

Author: I will keep it for now :)

|===

[[match-label-expressions]]
=== MATCH using label expressions

.Node pattern using the `OR` (`|`) label expression
[source, cypher]
----
MATCH (n:Movie|Person)
RETURN n.name AS name, n.title AS title
----

.Result
[role="queryresult",options="header,footer",cols="2*<m"]
|===
| name | title
| "Charlie Sheen" | <null>
| "Martin Sheen" | <null>
| "Michael Douglas" | <null>
| "Oliver Stone" | <null>
| "Rob Reiner" | <null>
| <null> | "Wall Street"
| <null> | "The American President"
2+|Rows: 7
|===

.Node pattern using negation (`!`) label expression
[source, cypher]
----
MATCH (n:!Movie)
RETURN labels(n) AS label, count(n) AS labelCount
----

[NOTE]
The above query uses the xref:functions/list.adoc#functions-labels[`labels()`] and xref:functions/aggregating.adoc#functions-count[`count()`] functions.

.Result
[role="queryresult",options="header,footer",cols="2*<m"]
|===
| label | labelCount
| ["Person"] | 5
2+|Rows: 1
|===

> For a full list of all label expressions supported by Cypher, see xref:patterns/reference.adoc#label-expressions[Patterns -> Label expressions].

Reviewer: ```suggestion
For a full list of label expressions supported by Cypher, see xref:patterns/reference.adoc#label-expressions[Patterns -> Label expressions].
```
alternatively, "-full"

|===
| path
| (:Person {name: "Charlie Sheen"})-[:ACTED_IN {role: "Bud Fox"}]->(:Movie {title: "Wall Street"})<-[:DIRECTED]-(:Person {name: "Oliver Stone"})
| (:Person {name: "Martin Sheen"})-[:ACTED_IN {role: "Carl Fox"}]->(:Movie {title: "Wall Street"})<-[:DIRECTED]-(:Person {name: "Oliver Stone"})
| (:Person {name: "Michael Douglas"})-[:ACTED_IN {role: "Gordon Gekko"}]->(:Movie {title: "Wall Street"})<-[:DIRECTED]-(:Person {name: "Oliver Stone"})
1+d| Rows: 3
|===

For more information about how `MATCH` is used to find patterns of varying complexity (including xref:patterns/variable-length-patterns.adoc#quantified-path-patterns[quantified path patterns], xref:patterns/variable-length-patterns.adoc#quantified-relationships[quantified relationships], and the xref:patterns/shortest-paths.adoc[shortest paths] between nodes), see the section on xref::patterns/index.adoc[Patterns].

== Multiple MATCH clauses, the WITH clause, and clause composition

In Cypher, the behavior of a query is defined by its clauses.
Each clause takes the current graph state and a table of intermediate results, processes them, and passes the updated graph state and results to the next clause.

> The first clause starts with the graph's initial state and an empty table, while the final clause produces the query's result.

Reviewer: ```suggestion
The first clause starts with the graph's initial state and an empty table, while the final clause produces the query result.
```

2+|Rows: 2
|===

[[match-on-relationship-type]]
=== Filter on relationship types

It is possible to specify the type of a relationship in a relationship pattern by using a colon (`:`) before the relationship type.

.Relationship pattern filtering on the `ACTED_IN` relationship type
[source, cypher]
----
MATCH (:Movie {title: 'Wall Street'})<-[:ACTED_IN]-(actor:Person)
RETURN actor.name AS actor
----

.Result
[source, role="queryresult",options="header,footer",cols="1*<m"]
|===
| actor
| "Michael Douglas"
| "Martin Sheen"
| "Charlie Sheen"
|Rows: 3
|===

[[match-on-multiple-relationship-types]]

> === Filter on multiple relationship types

Reviewer: for nodes, the new heading is "MATCH using label expressions"
makes me wonder if the two could be unified, "MATCH using node label expressions" and "MATCH using relationship label expressions" or something similar

Author: Yep - will do relationship type expressions

> It is possible to match for a pattern containing one out of several relationship types using the `OR` symbol, `|`.

Reviewer: ```suggestion
It is possible to match a pattern containing one of several relationship types using the `OR` symbol, `|`.
```

.Relationship pattern including either `ACTED_IN` or `DIRECTED` relationship types
[source, cypher]
----
MATCH (:Movie {title: 'Wall Street'})<-[:ACTED_IN|DIRECTED]-(person:Person)
RETURN person.name AS person
----

.Result
[role="queryresult",options="header,footer",cols="1*<m"]
|===
| person
| "Oliver Stone"
| "Michael Douglas"
| "Martin Sheen"
| "Charlie Sheen"
|Rows: 4
|===

As relationships have exactly one type each, `()-[:A&B]->()` will never match a relationship.

[[multiple-relationships]]
=== Find multiple relationships

A graph pattern can contain several relationship patterns.

.Graph pattern including several relationship patterns
[source, cypher]
----
MATCH (:Person {name: 'Charlie Sheen'})-[:ACTED_IN]->(movie:Movie)<-[:DIRECTED]-(director:Person)
RETURN movie.title AS movieTitle, director.name AS director
----

.Result
[role="queryresult",options="header,footer",cols="2*<m"]
|===
| movieTitle | director
| "Wall Street" | "Oliver Stone"
2+|Rows: 1
|===

[[where-predicates]]
== MATCH with WHERE predicates

> The `MATCH` clause is often paired with a `WHERE` sub-clause, which adds predicates to refine the patterns, making them more specific.

Reviewer: on "sub-clause" - a quick look at the source files of the cypher manual reveals that we're not using "sub-clause" or "subclause" consistently (15x in 5 files for sub-clause, 7x in 7 files for subclause).

i guess it is sub-clause - but we should follow up with changing the occurrences of ""subclause".

Author: Ok, we can do that.

These predicates are part of the pattern itself, not just filters applied after matching.

> Thus, the `WHERE` clause should always be placed with its corresponding `MATCH` clause.

Reviewer: ```suggestion
Thus, always place the `WHERE` clause with its corresponding `MATCH` clause.
```

tried to avoid "should" and to shorten it with imperative

The `MATCH` clause can be used with parameters.

.Parameters
[source, parameters]
----
{
  "movieTitle": "Wall Street",
  "actorRole": "Fox"
}
----

.Find nodes using paramters
[source, cypher]
----
MATCH (:Movie {title: $movieTitle})<-[r:ACTED_IN]-(p:Person)
WHERE r.role CONTAINS $actorRole
RETURN p.name AS actor, r.role AS role
----

[NOTE]
====
Relationships will only be matched once inside a single pattern.
Read more about this in the section on xref::patterns/reference.adoc#graph-patterns-rules-relationship-uniqueness[relationship uniqueness].
====
The above query uses the xref:syntax/operators.adoc#query-operator-comparison-string-specific[`CONTAINS` operator].

.Result
[role="queryresult",options="header,footer",cols="2*<m"]
|===
| actor | role

[[rel-types-with-uncommon-chars]]
=== Relationship types with uncommon characters
| "Charlie Sheen" | "Bud Fox"
| "Martin Sheen"  | "Carl Fox"

Databases occasionally contain relationship types including non-alphanumerical characters, or with spaces in them.
These are created using backticks (```).
2+|Rows: 2
|===

For more information about how to set parameters, see xref:syntax/parameters.adoc[Syntax -> Parameters].

> For example, the following query creates a relationship which contains a space (`OLD FRIENDS`) between `Martin Sheen` and `Rob Reiner`.

Reviewer: the "OLD FRIENDS" example got dropped completely, is this intentional?

Author: Yeah - it's pretty niche and seemed better suited the CREATE page if anything.

> In Cypher, a query’s behavior is defined by its clauses.

Reviewer: ```suggestion
In Cypher, a query's behavior is defined by its clauses.
```
alternatively, of-construction instead of genitive s
(suggestion was about the tick character though. there might be more occurrences of this character)

.Chaining consecutive `MATCH` clauses
[source, cypher]
----
MATCH (:Person {name: 'Martin Sheen'})-[:ACTED_IN]->(movie:Movie) // <1>
MATCH (director:Person)-[:DIRECTED]->(movie) // <2>
RETURN director.name AS director, movie.title AS movieTitle
----
<1> The result of the first `MATCH` clause is the variable `movie` which holds all the `Movies` that `Martin Sheen` has `ACTED_IN`.
<2> The second `MATCH` clause uses the `movie` variable to find any `Person` node with a `DIRECTED` relationship to those `Movie` nodes that `Martin Sheen` has `ACTED_IN`.

.Result
[role="queryresult",options="header,footer",cols="2*<m"]
|===
| director | movieTitle

| "Oliver Stone" | "Wall Street"
| "Rob Reiner"   | "The American President"

2+d| Rows: 2
|===

A variable can be implicitly carried over to the following clause by being referenced in another operation.
A variable can also be explicitly passed to the following clause using the xref:clauses/with.adoc[`WITH`] clause.

> If a variable is neither implicitly nor explicitly carried over to its following clause, it will be discarded and unavailable for reference later in the query.

Reviewer: ```suggestion
If a variable is neither implicitly nor explicitly carried over to its following clause, it will be discarded and is not available for reference later in the query.
```

.Using `WITH` and multiple `MATCH` clauses
[source, cypher]
----
MATCH (actors:Person)-[:ACTED_IN]->(movies:Movie) // <1>
WITH actors, count(movies) AS movieCount // <2>
ORDER BY movieCount DESC
LIMIT 1 // <3>
MATCH (actors)-[:ACTED_IN]->(movies) // <4>
RETURN actors.name AS actor, movieCount, collect(movies.title) AS movies
----

> <1> The `Person` and `Movie` nodes matched in this step are stored in variables.
> which are then passed on to the second row of the query.

Reviewer: ```suggestion
<1> The `Person` and `Movie` nodes matched in this step are stored in variables, which are then passed on to the second row of the query.
```

> <2> The `movie` variable is here implicitly imported by its occurrence in the `count()` function.

Reviewer 2: ```suggestion
<2> The `movies` variable here is implicitly imported by its occurrence in the `count()` function.
```

> <2> The `movies` variable is here implicitly imported by its occurrence in the `count()` function.

Reviewer: ```suggestion
<2> The `movies` variable is implicitly imported by its occurrence in the `count()` function.
```

The `WITH` clause explicitly imports the `actors` variable.

> <3> An xref:clauses/order-by.adoc[`ORDER BY`]  clause orders the results by `movieCount` in descending order, ensuring that the `Person` with the highest number of movies appears at the top, and xref:clauses/limit.adoc[`LIMIT] 1` ensures that all other `Person` nodes are discarded.

Reviewer: ```suggestion
<3> An xref:clauses/order-by.adoc[`ORDER BY`] clause orders the results by `movieCount` in descending order, ensuring that the `Person` with the highest number of movies appears at the top, and xref:clauses/limit.adoc[`LIMIT] 1` ensures that all other `Person` nodes are discarded.
```

> <4> The second MATCH clause finds all `Movie` nodes associated with the `Person` nodes currently bound to the `actors` variable.

Reviewer 2: ```suggestion
<4> The second `MATCH` clause finds all `Movie` nodes associated with the `Person` nodes currently bound to the `actors` variable.
```
